merlin: add new contexts commands

jchavarri opened the pull request:
Fixes #10222.

This work is part of the multi-context support for vscode (see https://github.com/ocamllabs/vscode-ocaml-platform/issues/1432).

## Current version

This PR prepares Dune to support multi-context config retrieval from Merlin by doing the following:
- Add an optional flag `--context` to `dune ocaml-merlin`: if passed, this flag will be used to return data to subsequent requests to the `ocaml-merlin` server
- Add `dune describe contexts` subcommands, which returns the list of Dune contexts available (1 per line)
- It also adds a new field `generate_merlin_rules` to the contexts declared in the workspace (originally implemented in #10328). This new field allows to generate merlin rules in any case, even if the context is not selected using `(merlin)`. This addition is useful to dynamically change the context for Merlin via the `SetContext` command mentioned above.

---

## Original version (now obsolete)

This PR adds two new commands to the ocaml-merlin API:

- `GetContexts`: returns the list of contexts available in the workspace (only ctxt names)
- `SetContext(name)`: sets the current context to get merlin data from

It also adds a new field `generate_merlin_rules` to the contexts declared in the workspace (originally implemented in #10328). This new field allows to generate merlin rules in any case, even if the context is not selected using `(merlin)`. This addition is useful to dynamically change the context for Merlin via the `SetContext` command mentioned above.

jchavarri committed:
merlin: add GetContexts command

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: add SetContexts (wip)

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
better tests

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: add rules regardless of (merlin)

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: update tests

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: simpler get-set-contexts test

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: introduce generate_merlin_rules

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
add docs and changelog

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
Revert "merlin: add rules regardless of (merlin)"

This reverts commit 3d3c97ff3d4aa231c55ce02528f1a5c39dd50752.

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: roll back changes in default-based-context test

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: update get-set-contexts test

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
update changes

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
rename Standard to Default

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: rename Nothing to Not_selected

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: fix tests

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri commented:
@voodoos Thanks for the feedback. I wasn't aware of how ocaml-lsp and Dune communicated through Dune's RPC feature, I understand you are referring to [this `Dune` module](https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server/src/dune.mli)?

I am not sure I understand how the suggested approach would work in terms of storing the information about the selected context. In the current implementation, the `SetContext` command will change the value of the current context:

https://github.com/ocaml/dune/blob/7a5395eb621a01e700ad2bab7897a446f527383d/bin/ocaml/ocaml_merlin.ml#L215

which is stored as a `ref` that is initialized as soon as `dune ocaml-merlin` starts:

https://github.com/ocaml/dune/blob/7a5395eb621a01e700ad2bab7897a446f527383d/bin/ocaml/ocaml_merlin.ml#L181

This is handy because this value can also be used by `dune ocaml-merlin` where the editor information is requested:

https://github.com/ocaml/dune/blob/7a5395eb621a01e700ad2bab7897a446f527383d/bin/ocaml/ocaml_merlin.ml#L186

As can be seen, this state handling happens only a few lines apart and doesn't spread to other modules outside `Ocaml_merlin`.

If the new commands were moved to Dune's RPC, then we would need to find a way to sync between the two processes (`rpc` and `ocaml-merlin`), because we would be writing the state through Dune's RPC, but we would still have to read it from the `ocaml-merlin` process. Or is there an alternative to avoid that?

> Adding new features to `dune ocaml-merlin` would only make a transition to the RPC more complex.

Considering the above, I think this addition doesn't change much the current state of things, and moving all the Merlin configuration handling and state at once to the RPC will be easier than spreading state across processes. I would be happy to eventually work on transitioning everything to the RPC if/when time permits, but that sounds like a much larger task and scope than the current goal: to allow users to select the context used by `print_merlin_conf`.

jchavarri commented:
> Since we might switch every plugins to using lsp in the future, and you seem to mostly target vscode, that might not be an issue ?

Unfortunately, I would need to support users that don't use vscode / ocaml-lsp and query merlin directly via vim or emacs.

voodoos commented on test/blackbox-tests/test-cases/merlin/get-set-contexts.t:
```diff
@@ -0,0 +1,86 @@
+Request config for file in alt context before calling SetContext
+
+  $ printf '(4:File%d:%s)' ${#FILE2} $FILE2 | dune ocaml-merlin | grep -i "$lib2" | sed 's/^[^:]*:[^:]*://'
+  No config found for file bar.ml. Try calling 'dune build'.))
```
I think we should improve the error messages now that the user might be using the wrong context since `dune build` won't solve that issue.

jchavarri commented:
@voodoos Thanks for the review. I wonder what your thoughts are about the suggestion in https://github.com/ocamllabs/vscode-ocaml-platform/issues/1432#issuecomment-2034676788? In particular, this part:

> add `--context CONTEXT` option `dune ocaml ocaml-merlin`

Maybe that would make easier to transition to RPC in the future? We would just need to move the `--context` flag from the `dune ocaml-merlin` executable to the RPC one. Besides that, I find interesting about the suggestion that it removes the need to patch `dot-merlin-reader`, which I was preparing in https://github.com/ocaml/merlin/pull/1743/.

jchavarri committed:
describe: add contexts subcommand

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: replace get/set context with flag

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: cleanup

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri commented:
@voodoos I have updated the PR to follow our recent decision to use a `--context` flag rather than a new protocol command to set the context. I also added a new subcommand `dune describe contexts` to list the contexts, which I will use from ocaml-lsp later on to offer clients the list of contexts available (to show in e.g. vscode a menu to choose from).

The work on [ocaml-lsp](https://github.com/ocaml/ocaml-lsp/pull/1238) is still ongoing, but any feedback you might have on this one is already much appreciated. Thanks!

jchavarri committed:
cleanup

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

voodoos commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -129,48 +133,71 @@ end = struct
+  let start ~ctx_name () =
+    let open Fiber.O in
+    let* setup = Import.Main.setup () in
+    let* setup = Memo.run setup in
```
I am wondering about the cost of these additional initializations. Is it really useful to check the existence of the context ? If it does not exist a "No config found for file %s." would be printed anyway.

jchavarri replied:
As it was a one-time thing (this code will run once when the ocaml-merlin process starts) I assumed it wouldn't have a lot of impact. Removed in https://github.com/ocaml/dune/pull/10324/commits/f21fbf837d177bedb8d332d236b8d2e97ef8a720.

jchavarri committed:
merlin: remove context check

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri commented:
The downstream PRs that depend on this feature are ready for review:

- [ocaml-lsp#1238](https://github.com/ocaml/ocaml-lsp/pull/1238/)
- [vscode-ocaml-platform#1449](https://github.com/ocamllabs/vscode-ocaml-platform/pull/1449)

The latter includes a short screen capture that helps visualize what the features added in this PR enable, from the user perspective.

jchavarri commented:
@rgrinberg @voodoos Apologies for pinging, I wonder if you have any thoughts on this PR? Thanks.

anmonteiro commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -129,48 +133,61 @@ end = struct
+      match ctx_name with
+      | None -> Default
+      | Some ctx_name ->
+        (match Context_name.of_string_opt ctx_name with
```
I was wondering what the valid names for contexts were and incidentally found a crash: https://github.com/ocaml/dune/pull/10472

anmonteiro commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -129,48 +133,61 @@ end = struct
+      | None -> Default
+      | Some ctx_name ->
+        (match Context_name.of_string_opt ctx_name with
+         | None -> User_error.raise [ Pp.textf "Invalid context name %S" ctx_name ]
```
we could provide a hint (maybe it could even live in `context_name.ml`) about what's a valid context name here.

jchavarri replied:
i agree. maybe this can be done in a separate PR (worth including a fix for your repro in #10472 as well).

anmonteiro commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -129,48 +133,61 @@ end = struct
 
   let dump_dot_merlin s =
-    to_local s
+    to_local ~selected_context:Default s
```
wouldn't this also benefit from printing the config for a given context argument?

jchavarri replied:
yup, added in https://github.com/ocaml/dune/pull/10324/commits/eb09dbaa6a1539822bcc0a06d650925590867d2b.

anmonteiro commented on doc/changes/10324.md:
```diff
@@ -0,0 +1,5 @@
+- Add new flag `--context` to `ocaml-merlin`, which allows to select a Dune
```
```suggestion
- Add new flag `--context` to `dune ocaml-merlin`, which allows to select a Dune
```

anmonteiro commented on doc/reference/dune-workspace/context.rst:
```diff
@@ -21,6 +21,9 @@ the description of an opam switch, as follows:
 - ``(merlin)`` instructs Dune to use this build context for Merlin.
 
+- ``(generate_merlin_rules)`` instructs Dune to generate Merlin rules for this
+  context, even if it is not the one selected for Merlin.
```
```suggestion
  context, even if it is not the one selected via ``(merlin)``.
```

anmonteiro commented on src/dune_rules/workspace.ml:
```diff
@@ -384,7 +406,8 @@ module Context = struct
       and+ loc = loc
-      and+ merlin = field_b "merlin" in
+      and+ merlin = field_b "merlin"
+      and+ generate_merlin_rules = field_b "generate_merlin_rules" in
```
should this have a `since 3.16` check?

jchavarri replied:
👍 done in https://github.com/ocaml/dune/pull/10324/commits/d384104f3f2872b1c16c2f5a61c556292675e190.

anmonteiro reviewed (commented):
This looks good overall and my suggestions are mostly minor. The important ones are:

- should we gate this on `lang dune 3.16`?
- should we make the other commands in `dune ocaml-merlin` take `--context` too, not just `print-conf`?

jchavarri committed:
apply suggestions from code review

Co-authored-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
gate generate_merlin_rules to 3.16

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: add --context to dump_dot_merlin

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri commented:
@anmonteiro thanks for the review!

>     * should we gate this on `lang dune 3.16`?
>     * should we make the other commands in `dune ocaml-merlin` take `--context` too, not just `print-conf`?

I thought the answer to these 2 questions is yes, so just included them in the PR, pls lmk if I'm missing anything :)

anmonteiro commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -264,6 +291,14 @@ module Dump_dot_merlin = struct
             ~doc:
               "The path to the folder of which the configuration should be printed. \
                Defaults to the current directory.")
+    and+ ctx_name =
```
I missed this at first, but I think this is where you should be parsing the context, instead of `arg_to_ctx`.

```ocaml
Arg.(value & opt Selected_context.conv Selected_context.Default & ...)
```

where `Selected_context.conv` is 


```ocaml
let conv =
  let parse ctx_name =
    match Context_name.of_string_opt ctx_name with
    | None -> User_error.raise [ Pp.textf "Invalid context name %S" ctx_name ]
    | Some ctx_name -> Custom ctx_name
  in
  let print = function Default -> "default" | Custom ctx -> ctx in
  Arg.conv ~docv:"context" (parse, print)
```

jchavarri replied:
Added in https://github.com/ocaml/dune/pull/10324/commits/666e5250015e19f5f5d391b7e35ac64e13e6963b.

anmonteiro commented on bin/ocaml/ocaml_merlin.ml:
```diff
@@ -264,6 +291,14 @@ module Dump_dot_merlin = struct
+        & info
+            [ "context" ]
+            ~docv:"CONTEXT"
+            ~doc:"The Dune context in which the command will return information for")
```
What do you think about this phrasing? I can't exactly point out what, but the current sentence reads sort of weird.

```suggestion
            ~doc:"Dune will return information for this build context")
```

anmonteiro reviewed (approved):
looks good. I just think we need to improve the context arg parsing. I made a suggestion that avoids creating the intermediate option type.

jchavarri committed:
apply suggestions from code review

Co-authored-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

jchavarri committed:
merlin: add Select_context.conv

Signed-off-by: Javier Chávarri <javier.chavarri@gmail.com>

anmonteiro commented:
I tried resolving conflicts on this but `ls -a ./_build/cross/.merlin-conf` and `ls -a ./_build/default/.merlin-conf` don't seem to exist anymore in tests (`test/blackbox-tests/test-cases/merlin/default-based-context.t/run.t`).

I wonder if it's because of #8567. Bisecting....

anmonteiro committed:
fix: promote tests after rebase

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented:
nevermind, I just messed up the rebase.

anmonteiro committed:
refactor: remove `Selected_context.t`, reuse the context arg, use it in dump

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented:
I also just pushed a commit that:

- removes `Selected_context.t` (`Context_name.t` and associated functions have all we need)
- add the `--context` arg to `dump` too
- reuse the context arg as `Selected_context.arg`

anmonteiro commented:
Thank you, let's do it
